conventions/deprecation.rst

```diff
@@ -65,3 +65,52 @@ You can simply mark a method defined inline in a class with
         // This failed to compile with GCC 3.3.5.
         XAPIAN_DEPRECATED(int old_inline_method()) { return 42; }
     };
+
+Implementing Deprecation Warnings for the Bindings
+--------------------------------------------------
+
+Currently we don't have an equivalent of the C++ ``XAPIAN_DEPRECATED()`` macro
+for the bindings, but it would be good to have.  Here are some notes on how
+this could be achieved for various languages we support:
+
+* PHP now has a ``E_USER_DEPRECATED`` error level - in a deprecated method we
+  could do:
+
+  .. code-block:: php
+
+     trigger_error(
+         'World::hi() is deprecated, use World::hello() instead',
+         XAPIAN_DEPRECATED
+     );
+
+* Python has ``DeprecationWarning``, which we were using in 1.2.x a bit:
+
+  .. code-block:: python
+
+     warnings.warn(
+         'World::hi() is deprecated, use World::hello() instead',
+         DeprecationWarning,
+     )
+
+* Ruby - there are external libraries to handle deprecation warnings, but the
+  simplest option without external dependencies seems to be:
+
+  .. code-block:: ruby
+
+     warn "[DEPRECATION] World::hi() is deprecated, use World::hello() instead"
+
+* Perl:
+
+  .. code-block:: perl
+
+     use warnings;
+     warnings::warnif('deprecated', 'World::hi() is deprecated, use World::hello() instead');
+
+* Java has ``@Deprecated``, but I think that's a documentation thing only.
```

Comment:
Actually, I was confusing `@deprecated` and `@Deprecated` here I think - the latter is an annotation and "ensures" the compiler will issue a warning - the former is javadoc and may give a warning with some compilers: https://docs.oracle.com/javase/7/docs/technotes/guides/javadoc/deprecation/deprecation.html#annotation